libp2p/protocols/connectivity/autonatv2/server.nim

```diff
@@ -0,0 +1,248 @@
+    if t.handles(addrs):
+      return true
+  return false
+
+proc chooseDialAddr(
+    self: AutonatV2, pid: PeerId, addrs: seq[MultiAddress]
+): Future[Opt[(Connection, AddrIdx)]] {.async: (raises: [CancelledError]).} =
+  for i, ma in addrs:
+    if self.canDial(ma):
+      try:
+        debug "Trying to dial", chosenAddrs = ma, addrIdx = i
+        return Opt.some(
+          (await self.switch.dial(pid, @[ma], $AutonatV2Codec.DialBack), i.AddrIdx)
+        )
+      except DialFailedError:
+        return Opt.none((Connection, AddrIdx))
+      except LPStreamError:
+        return Opt.none((Connection, AddrIdx))
+  return Opt.none((Connection, AddrIdx))
+
+proc handleDialRequest(
+    self: AutonatV2, conn: Connection, req: DialRequest
+) {.async: (raises: [CancelledError, DialFailedError, LPStreamError]).} =
+  let observedIPAddr = (await conn.findObservedIPAddr(req)).valueOr:
+    error "Could not find observed IP address"
+    return
+
+  let (dialBackConn, addrIdx) = (await self.chooseDialAddr(conn.peerId, req.addrs)).valueOr:
+    error "No dialable addresses found"
+    await conn.sendDialResponse(ResponseStatus.EDialRefused)
+    return
+  defer:
+    await dialBackConn.close()
+
+  # if observed address for peer is not in address list to try
+  # then we perform Amplification Attack Prevention
+  if not ipAddrMatches(observedIPAddr, req.addrs):
+    debug "Starting amplification attack prevention",
+      observedIPAddr = observedIPAddr, testAddr = req.addrs[addrIdx]
+    # send DialDataRequest and wait until dataReceived is enough
+    if not await self.amplificationAttackPrevention(conn, addrIdx):
+      return
+
+  debug "Sending DialBack",
+    nonce = req.nonce, addrIdx = addrIdx, addr = req.addrs[addrIdx]
+
+  let dialStatus = await dialBackConn.dialBack(req.nonce)
+
+  await conn.sendDialResponse(
+    ResponseStatus.Ok, addrIdx = Opt.some(addrIdx), dialStatus = Opt.some(dialStatus)
+  )
+
+proc new*(
+    T: typedesc[AutonatV2],
+    switch: Switch,
+    dialTimeout: Duration = DefaultDialTimeout,
+    dialDataSize: uint64 = DefaultDialDataSize,
+    amplificationAttackTimeout: Duration = DefaultAmplificationAttackDialTimeout,
+): T =
+  let autonatV2 = T(
+    switch: switch,
+    dialTimeout: dialTimeout,
+    dialDataSize: dialDataSize,
+    amplificationAttackTimeout: amplificationAttackTimeout,
+  )
+  proc handleStream(
+      conn: Connection, proto: string
+  ) {.async: (raises: [CancelledError]).} =
+    try:
+      let msg = AutonatV2Msg.decode(
+        initProtoBuffer(await conn.readLp(AutonatV2MsgLpSize))
+      ).valueOr:
+        raise newException(AutonatV2Error, "Unable to decode AutonatV2Msg")
+      debug "Received message", msgType = $msg.msgType
+      if msg.msgType != MsgType.DialRequest:
+        raise newException(AutonatV2Error, "Expecting DialRequest, got " & $msg.msgType)
+      await autonatV2.handleDialRequest(conn, msg.dialReq)
+    except LPStreamRemoteClosedError as exc:
```

Comment:
would be better if this `try` block is not catching error that are thrown from this code here

Reply:
move msg deserialization and msg type check above try?

Reply:
initializes rawMsg first and handling only connection errors here (readLp call)